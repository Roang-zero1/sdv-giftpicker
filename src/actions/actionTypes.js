export const INCREMENT_ITEM_COUNT = 'INCREMENT_ITEM_COUNT';
export const SET_LOADED = 'SET_LOADED';
export const UPDATE_ITEMS = 'UPDATE_ITEMS';
export const SET_PROGRESS = 'SET_PROGRESS';
export const SET_GIFT_COUNT = 'SET_GIFT_COUNT';
