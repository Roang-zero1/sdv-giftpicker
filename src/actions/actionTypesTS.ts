enum ActionTypes {
  UPDATE_ITEMS = 'items/UPDATE_ITEMS',
  MODIFY_COUNT = 'items/MODIFY_COUNT'
}

export default ActionTypes;
