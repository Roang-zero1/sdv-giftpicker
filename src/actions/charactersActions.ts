import { createStandardAction } from 'typesafe-actions';
import { ICharacterGiftCount, ICharacterGiftSelect } from '../common/types';

export const setGiftCount = createStandardAction('characters/SET_GIFT_COUNT')<
  ICharacterGiftCount
>();
export const selectGift = createStandardAction('characters/SELECT_GIFT')<
  ICharacterGiftSelect
>();
export const deselectGift = createStandardAction('characters/DESELECT_GIFT')<
  ICharacterGiftSelect
>();
