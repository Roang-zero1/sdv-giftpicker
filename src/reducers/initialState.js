export default {
  characters: {},
  status: { intro: false, save: false }
};
