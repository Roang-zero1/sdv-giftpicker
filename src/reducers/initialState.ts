import { initialState as initialCharactersState } from './charactersReducer';
import { initialState as initialItemsState } from './itemsReducer';
import { initialState as initialNavigationState } from './navigationReducer';
import { initialState as initialStatusState } from './statusReducer';

export default {
  characters: initialCharactersState,
  items: initialItemsState,
  navigation: initialNavigationState,
  status: initialStatusState
};
