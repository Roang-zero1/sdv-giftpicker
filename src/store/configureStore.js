import { persistReducer, persistStore } from 'redux-persist';

import { createStore } from 'redux';
import rootReducer from '../reducers/rootReducer';
import storage from 'redux-persist/lib/storage';

const persistConfig = {
  blacklist: ['navigation'],
  key: 'sdv-gp',
  storage
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = createStore(
  persistedReducer,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
);
export const persistor = persistStore(store);
