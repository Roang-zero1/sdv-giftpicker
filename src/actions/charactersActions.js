import * as types from './actionTypes';

export function setGiftCount(character, count) {
  return {
    char: character,
    count: count,
    type: types.SET_GIFT_COUNT
  };
}

export function selectGift(character, itemID) {
  return {
    char: character,
    gift: itemID,
    type: types.SELECT_GIFT
  };
}

export function deselectGift(character, itemID) {
  return {
    char: character,
    gift: itemID,
    type: types.DESELECT_GIFT
  };
}
