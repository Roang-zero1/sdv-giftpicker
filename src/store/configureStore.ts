import { PersistConfig, persistReducer, persistStore } from 'redux-persist';

import { createStore, StoreEnhancer } from 'redux';
import storage from 'redux-persist/lib/storage';
import initialState from '../reducers/initialState';
import rootReducer from '../reducers/rootReducer';

const persistConfig: PersistConfig = {
  blacklist: ['navigation'],
  key: 'sdv-gp',
  storage
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const devTools: StoreEnhancer = (window as any).__REDUX_DEVTOOLS_EXTENSION__
  ? (window as any).__REDUX_DEVTOOLS_EXTENSION__()
  : (f: any) => f;

// @ts-ignore Currently bugged for redux 4.0.0
export const store = createStore(persistedReducer, initialState, devTools);
export const persistor = persistStore(store);
