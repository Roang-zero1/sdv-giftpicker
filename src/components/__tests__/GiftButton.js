import 'jest-styled-components';

import ConnectedGiftButton, { GiftButton } from '../GiftButton';
import { mount, shallow } from 'enzyme';

import React from 'react';
import configureStore from 'redux-mock-store';
import giftsData from '../../data/GiftsData.json';
import initialState from '../../reducers/initialState';
import toJson from 'enzyme-to-json';
import * as charactersActions from '../../actions/charactersActions';
import { getType } from 'typesafe-actions';

const giftID = 20;
const charName = 'Lewis';

const defaultProps = {
  characters: { Lewis: {} },
  items: {},
  status: { save: true }
};

describe('components/GiftButton --- Shallow render connected component', () => {
  let renderedCut;
  let props;

  const cut = () => {
    if (!renderedCut) {
      renderedCut = shallow(
        <GiftButton gift={20} char={charName} {...props} />
      );
    }
    return renderedCut;
  };

  beforeEach(() => {
    renderedCut = undefined;
    props = defaultProps;
  });

  it('should render the component', () => {
    expect(cut()).toHaveLength(1);
    expect(cut().find('Gift')).toHaveLength(1);
    expect(cut().find('StyledButton')).toHaveLength(1);
    expect(cut().find('GiftText')).toHaveLength(1);
    expect(cut().find('GiftIcon')).toHaveLength(1);
  });

  it('should render the component with a count', () => {
    expect(cut().find('GiftCount')).toHaveLength(1);
  });

  it('should render the component without a count', () => {
    props = {
      ...props,
      status: { save: false }
    };
    expect(cut().find('GiftCount')).toHaveLength(0);
  });

  it('should change the color for selected gifts', () => {
    props = {
      ...props,
      characters: { Lewis: { selected: [20] } }
    };
    expect(
      cut()
        .find('StyledButton')
        .prop('color')
    ).toEqual('success');
  });

  it('should be the same as the last snapshot', () => {
    expect(toJson(cut())).toMatchSnapshot();
  });
});

describe('components/GiftButton --- Render connected component', () => {
  let renderedCut;
  let props;

  const cut = () => {
    if (!renderedCut) {
      renderedCut = mount(
        <GiftButton gift={giftID} char={charName} {...props} />
      );
    }
    return renderedCut;
  };

  beforeEach(() => {
    renderedCut = undefined;
    props = defaultProps;
  });

  it('should render the gift name', () => {
    expect(
      cut()
        .find('GiftText')
        .text()
    ).toEqual(giftsData[giftID].displayName);
  });

  it('should render the component with a count of zero', () => {
    expect(
      cut()
        .find('GiftCount')
        .text()
    ).toEqual('0');
  });

  it('should render the component with a count of zero if there is a negative count', () => {
    let count = -2;
    props = {
      ...props,
      items: {
        20: count
      }
    };
    expect(
      cut()
        .find('GiftCount')
        .text()
    ).toEqual(`0`);
  });

  it('should render the component with the correct count', () => {
    let count = 120;
    props = {
      ...props,
      items: {
        20: count
      }
    };
    expect(
      cut()
        .find('GiftCount')
        .text()
    ).toEqual(`${count}`);
  });

  it('should add an css order if the item is not in the current items', () => {
    expect(
      toJson(
        cut()
          .find('Gift')
          .children()
          .first()
      )
    ).toHaveStyleRule('order', '1');
  });

  it('should add an css order if the item has a count of 0', () => {
    props = {
      ...props,
      items: {
        '20': 0
      }
    };
    expect(
      toJson(
        cut()
          .find('Gift')
          .children()
          .first()
      )
    ).toHaveStyleRule('order', '1');
  });

  it('should be the same as the last snapshot', () => {
    expect(toJson(cut())).toMatchSnapshot();
  });
});

describe('components/GiftButton --- Connected component tests', () => {
  const mockStore = configureStore();
  let renderedGiftButton;
  let store;
  let props;

  const cut = () => {
    const state = {
      ...initialState,
      ...defaultProps
    };
    store = mockStore(state);
    if (!renderedGiftButton) {
      renderedGiftButton = mount(
        <ConnectedGiftButton
          gift={giftID}
          char={charName}
          store={store}
          {...props}
        />
      );
    }
    return renderedGiftButton;
  };

  beforeEach(() => {
    renderedGiftButton = undefined;
    props = {};
  });

  it('should add a gift on button press', () => {
    let button = cut().find('Button');
    button.simulate('click');
    let actions = store.getActions();
    expect(actions[0]).toEqual({
      payload: {
        char: charName,
        itemID: giftID
      },
      type: getType(charactersActions.selectGift)
    });
  });

  it('should remove a gift on button press', () => {
    props = {
      ...props,
      deselect: true
    };
    let button = cut().find('Button');
    button.simulate('click');
    let actions = store.getActions();
    expect(actions[0]).toEqual({
      payload: {
        char: charName,
        itemID: giftID
      },
      type: getType(charactersActions.deselectGift)
    });
  });
});
