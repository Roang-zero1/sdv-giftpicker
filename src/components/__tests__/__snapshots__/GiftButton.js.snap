// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/GiftButton --- Render connected component should be the same as the last snapshot 1`] = `
.c2 {
  left: -12px;
  position: relative;
  max-width: 24px;
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
}

.c0 {
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
  width: 100%;
}

.c1 div,
.c1 img {
  pointer-events: none;
}

.c3 {
  -webkit-flex: 1 1 2vw;
  -ms-flex: 1 1 2vw;
  flex: 1 1 2vw;
  overflow: hidden;
}

.c4 {
  text-align: right;
}

<GiftButton
  char="Lewis"
  characters={
    Object {
      "Lewis": Object {},
    }
  }
  deselect={false}
  gift={20}
  items={Object {}}
  status={
    Object {
      "save": true,
    }
  }
>
  <Gift
    className="mb-1"
    md="6"
    owned={false}
    xl="4"
    xs="12"
  >
    <Component
      className="mb-1 c0"
      md="6"
      owned={false}
      xl="4"
      xs="12"
    >
      <Col
        className="mb-1 c0"
        md="6"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
        xl="4"
        xs="12"
      >
        <div
          className="mb-1 c0 col-12 col-md-6 col-xl-4"
        >
          <StyledButton
            className="flex-nowrap ml-2 mr-2 row"
            color="dark"
            onClick={[Function]}
            outline={true}
          >
            <Button
              className="flex-nowrap ml-2 mr-2 row c1"
              color="dark"
              onClick={[Function]}
              outline={true}
              tag="button"
            >
              <button
                className="flex-nowrap ml-2 mr-2 row c1 btn btn-outline-dark"
                onClick={[Function]}
                type="button"
              >
                <Col
                  tag="div"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                  xs="1"
                >
                  <div
                    className="col-1"
                  >
                    <Icon
                      gift={20}
                      grayscale={true}
                    >
                      <styled.img
                        alt=""
                        grayscale={true}
                        src="Leek.png"
                      >
                        <img
                          alt=""
                          className="c2"
                          src="Leek.png"
                        />
                      </styled.img>
                    </Icon>
                  </div>
                </Col>
                <GiftText>
                  <Col
                    className="c3"
                    tag="div"
                    widths={
                      Array [
                        "xs",
                        "sm",
                        "md",
                        "lg",
                        "xl",
                      ]
                    }
                  >
                    <div
                      className="c3 col"
                    >
                      Leek
                    </div>
                  </Col>
                </GiftText>
                <GiftCount
                  xs="3"
                >
                  <Col
                    className="c4"
                    tag="div"
                    widths={
                      Array [
                        "xs",
                        "sm",
                        "md",
                        "lg",
                        "xl",
                      ]
                    }
                    xs="3"
                  >
                    <div
                      className="c4 col-3"
                    >
                      0
                    </div>
                  </Col>
                </GiftCount>
              </button>
            </Button>
          </StyledButton>
        </div>
      </Col>
    </Component>
  </Gift>
</GiftButton>
`;

exports[`components/GiftButton --- Shallow render connected component should be the same as the last snapshot 1`] = `
<Gift
  className="mb-1"
  md="6"
  owned={false}
  xl="4"
  xs="12"
>
  <StyledButton
    className="flex-nowrap ml-2 mr-2 row"
    color="dark"
    onClick={[Function]}
    outline={true}
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
      xs="1"
    >
      <Icon
        gift={20}
        grayscale={true}
      />
    </Col>
    <GiftText>
      Leek
    </GiftText>
    <GiftCount
      xs="3"
    >
      0
    </GiftCount>
  </StyledButton>
</Gift>
`;
